tests/python/test_uds.py

```diff
@@ -0,0 +1,119 @@
+#!/usr/bin/env python3
+from parameterized import parameterized_class
+import unittest
+import struct
+import threading
+
+from panda.python.uds import SERVICE_TYPE, DATA_IDENTIFIER_TYPE, IsoTpMessage, UdsClient, get_rx_addr_for_tx_addr
+
+DEFAULT_VIN = '1H3110W0RLD5'
+
+
+class UdsServer(UdsClient):
+  def __init__(self, *args, **kwargs):
+    super().__init__(*args, **kwargs, single_frame_mode=True)
+    self.kill_event = threading.Event()
+    self.uds_thread = threading.Thread(target=self._uds_response)
+
+  def start(self):
+    if self.uds_thread.is_alive():
+      self.stop()
+    self.uds_thread.start()
+
+  def stop(self):
+    self.kill_event.set()
+    self.uds_thread.join()
+
+  def _uds_response(self):
+    # send request, wait for response
+    max_len = 8 if self.sub_addr is None else 7
+    isotp_msg = IsoTpMessage(self._server_can_client, timeout=0, debug=self.debug, max_len=max_len, single_frame_mode=True)
+    isotp_msg.send(b"", setup_only=True)
+    while not self.kill_event.is_set():
+      resp, _ = isotp_msg.recv(0)
+      print('UDS Server - got resp', resp)
+
+      # message from client not fully built
+      if resp is None:
+        continue
+
+      # got a message, now respond
+      resp_sid = resp[0] if len(resp) > 0 else None
+
+      if resp_sid == SERVICE_TYPE.TESTER_PRESENT:
+        dat = bytes([SERVICE_TYPE.TESTER_PRESENT + 0x40, 0x0])
+        print('Car ECU - sending back:', dat)
+        isotp_msg.send(dat)
+      elif resp_sid == SERVICE_TYPE.READ_DATA_BY_IDENTIFIER:
+        print('Car ECU - READ DATA')
+        assert len(resp) > 1
+        resp_data = struct.unpack('!H', resp[1:])[0]
+        print(resp_data)
+        if resp_data == DATA_IDENTIFIER_TYPE.VIN:
+          dat = bytes([resp_sid + 0x40]) + resp[1:] + DEFAULT_VIN.encode('utf-8')
+          print('Car ECU - sending back:', dat)
+          isotp_msg.send(dat)
+        else:
+          dat = bytes([0x7F, resp_sid, *resp[1:]])
+          print('Car ECU - bad data id, sending back:', dat)
+          isotp_msg.send(dat)
```

Comment:
I would have a lookup table something like `service` -> `subfunction (can be empty string)` -> `remaining data` -> `response message` with a function used to add entries to this lookup table.  If the service isn't found it should return the error `service not supported` and if the subfunction isn't found it should return `sub-function not supported`, etc...